### PR TITLE
fix revalidation issue with route handlers

### PR DESCRIPTION
**Pull request by ztanner**
### What
When a route handler uses an API that opts it into dynamic rendering (such as `no-store` on a fetch), and also specifies a `revalidate` time, the `revalidate` time is ignored and route is treated as fully static.

### Why
`revalidate: 0` and `revalidate: false` have different semantic meanings: `false` essentially means cache forever, whereas `0` means it's dynamic. Since `0` is also falsey, the code we have to fallback with a default `revalidate` value for route handlers is incorrectly not marking the route as dynamic, and as a result, caching the route without an expiration time.

### How
This updates the fallback handling for app routes respect a revalidation value of `0`, so that the page can properly be marked dynamic.

### Test Explanation
This adds 2 new routes handlers: both have a revalidation time specified & use `no-store` on a fetch, but only one of them specifies `export const dynamic = 'force-static'`. The one that doesn't specify `force-static` is correctly omitted from the prerender manifest. The one that is `force-static` is correctly in the prerender manifest with the right expiration time. An additional test case was added to verify that this data refreshes after the specified interval.

Closes NEXT-2764

### CODE REVIEW

**Comment by ztanner** [comment]
* **#63213** <a href="https://app.graphite.dev/github/pr/vercel/next.js/63213?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `canary`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @ztanner and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Comment by ijjk** [comment]
# Stats from current PR

<details open>
<summary><strong>Default Build</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js 03-12-fix_revalidation_issue_with_route_handlers | Change |
| - | - | - | - |
| buildDuration | 14.1s | 14.1s | N/A |
| buildDurationCached | 7.7s | 6.6s | N/A |
| nodeModulesSize | 198 MB | 198 MB | ⚠️ +416 B |
| nextStartRea..uration (ms) | 436ms | 439ms | N/A |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js 03-12-fix_revalidation_issue_with_route_handlers | Change |
| - | - | - | - |
| 2453-HASH.js gzip | 30.5 kB | 30.5 kB | N/A |
| 3304.HASH.js gzip | 181 B | 181 B | ✓ |
| 3f784ff6-HASH.js gzip | 53.7 kB | 53.7 kB | N/A |
| 8299-HASH.js gzip | 5.04 kB | 5.04 kB | N/A |
| framework-HASH.js gzip | 45.2 kB | 45.2 kB | ✓ |
| main-app-HASH.js gzip | 242 B | 242 B | ✓ |
| main-HASH.js gzip | 32.2 kB | 32.2 kB | N/A |
| webpack-HASH.js gzip | 1.68 kB | 1.68 kB | N/A |
| Overall change | 45.6 kB | 45.6 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js 03-12-fix_revalidation_issue_with_route_handlers | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | vercel/next.js 03-12-fix_revalidation_issue_with_route_handlers | Change |
| - | - | - | - |
| _app-HASH.js gzip | 196 B | 197 B | N/A |
| _error-HASH.js gzip | 184 B | 184 B | ✓ |
| amp-HASH.js gzip | 505 B | 505 B | ✓ |
| css-HASH.js gzip | 324 B | 325 B | N/A |
| dynamic-HASH.js gzip | 2.5 kB | 2.5 kB | N/A |
| edge-ssr-HASH.js gzip | 258 B | 258 B | ✓ |
| head-HASH.js gzip | 352 B | 352 B | ✓ |
| hooks-HASH.js gzip | 370 B | 371 B | N/A |
| image-HASH.js gzip | 4.21 kB | 4.21 kB | ✓ |
| index-HASH.js gzip | 259 B | 259 B | ✓ |
| link-HASH.js gzip | 2.67 kB | 2.67 kB | N/A |
| routerDirect..HASH.js gzip | 314 B | 312 B | N/A |
| script-HASH.js gzip | 386 B | 386 B | ✓ |
| withRouter-HASH.js gzip | 309 B | 309 B | ✓ |
| 1afbb74e6ecf..834.css gzip | 106 B | 106 B | ✓ |
| Overall change | 6.57 kB | 6.57 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | vercel/next.js 03-12-fix_revalidation_issue_with_route_handlers | Change |
| - | - | - | - |
| _buildManifest.js gzip | 481 B | 484 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js 03-12-fix_revalidation_issue_with_route_handlers | Change |
| - | - | - | - |
| index.html gzip | 528 B | 529 B | N/A |
| link.html gzip | 541 B | 541 B | ✓ |
| withRouter.html gzip | 523 B | 523 B | ✓ |
| Overall change | 1.06 kB | 1.06 kB | ✓ |

</details>

<details>
<summary><strong>Edge SSR bundle Size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js 03-12-fix_revalidation_issue_with_route_handlers | Change |
| - | - | - | - |
| edge-ssr.js gzip | 95.1 kB | 95.1 kB | N/A |
| page.js gzip | 3.06 kB | 3.07 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Middleware size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js 03-12-fix_revalidation_issue_with_route_handlers | Change |
| - | - | - | - |
| middleware-b..fest.js gzip | 627 B | 622 B | N/A |
| middleware-r..fest.js gzip | 151 B | 151 B | ✓ |
| middleware.js gzip | 25.5 kB | 25.5 kB | N/A |
| edge-runtime..pack.js gzip | 839 B | 839 B | ✓ |
| Overall change | 990 B | 990 B | ✓ |

</details>

<details>
<summary><strong>Next Runtimes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js 03-12-fix_revalidation_issue_with_route_handlers | Change |
| - | - | - | - |
| app-page-exp...dev.js gzip | 171 kB | 171 kB | ✓ |
| app-page-exp..prod.js gzip | 96.6 kB | 96.6 kB | ✓ |
| app-page-tur..prod.js gzip | 98.3 kB | 98.3 kB | ✓ |
| app-page-tur..prod.js gzip | 92.8 kB | 92.8 kB | ✓ |
| app-page.run...dev.js gzip | 149 kB | 149 kB | ✓ |
| app-page.run..prod.js gzip | 91.3 kB | 91.3 kB | ✓ |
| app-route-ex...dev.js gzip | 21.3 kB | 21.3 kB | ✓ |
| app-route-ex..prod.js gzip | 15 kB | 15 kB | ✓ |
| app-route-tu..prod.js gzip | 15 kB | 15 kB | ✓ |
| app-route-tu..prod.js gzip | 14.8 kB | 14.8 kB | ✓ |
| app-route.ru...dev.js gzip | 21 kB | 21 kB | ✓ |
| app-route.ru..prod.js gzip | 14.8 kB | 14.8 kB | ✓ |
| pages-api-tu..prod.js gzip | 9.52 kB | 9.52 kB | ✓ |
| pages-api.ru...dev.js gzip | 9.8 kB | 9.8 kB | ✓ |
| pages-api.ru..prod.js gzip | 9.52 kB | 9.52 kB | ✓ |
| pages-turbo...prod.js gzip | 22.3 kB | 22.3 kB | ✓ |
| pages.runtim...dev.js gzip | 22.9 kB | 22.9 kB | ✓ |
| pages.runtim..prod.js gzip | 22.3 kB | 22.3 kB | ✓ |
| server.runti..prod.js gzip | 50.5 kB | 50.5 kB | ✓ |
| Overall change | 948 kB | 948 kB | ✓ |

</details>

<details>
<summary><strong>build cache</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js 03-12-fix_revalidation_issue_with_route_handlers | Change |
| - | - | - | - |
| 0.pack gzip | 1.56 MB | 1.56 MB | ⚠️ +125 B |
| index.pack gzip | 105 kB | 105 kB | N/A |
| Overall change | 1.56 MB | 1.56 MB | ⚠️ +125 B |

</details>

<details>
<summary><strong>Diff details</strong></summary>

<details>
<summary>Diff for <strong>middleware.js</strong></summary>

Diff too large to display
</details>

</details>

</details>
Commit: 60bea91abbbc58e7fdd24f2f7a7cb24123a68db0

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 60bea91abbbc58e7fdd24f2f7a7cb24123a68db0

`pnpm test-start test/e2e/app-dir/app-static/app-static-custom-handler.test.ts`
- app-dir static/dynamic handling > should properly revalidate a route handler that triggers dynamic usage with force-static

<details>
<summary>Expand output</summary>

  ● app-dir static/dynamic handling › should properly revalidate a route handler that triggers dynamic usage with force-static

    expect(received).not.toBe(expected) // Object.is equality

    Expected: not 1710281992675

      233 |       data = await res.json()
      234 |
    > 235 |       expect(data.now).not.toBe(initialTimestamp)
          |                            ^
      236 |     })
      237 |
      238 |     if (!process.env.CUSTOM_CACHE_HANDLER) {

      at Object.toBe (e2e/app-dir/app-static/app-static.test.ts:235:28)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Review by ijjk — APPROVED** [review]

File: packages/next/src/export/routes/app-route.ts [hunk 1]
```diff
@@ -91,7 +91,10 @@ export async function exportAppRoute(
     }
 
     const blob = await response.blob()
-    const revalidate = context.renderOpts.store?.revalidate || false
+    const revalidate =
+      typeof context.renderOpts.store?.revalidate === 'undefined'
+        ? false
+        : context.renderOpts.store.revalidate
```
Comment (wyattjoh) [inline packages/next/src/export/routes/app-route.ts:97]: Good use case for the `??` operator!

**Pull request merged**